File: C++/Min distance between two given nodes of a Binary Tree.cpp
class Solution{
    private:
    Node * fillParent(unordered_map<Node *,Node *> &parent,Node *root,int a){
        Node * start = NULL;
        queue<Node *> q;
        q.push(root);
        while(!q.empty()){
            Node * node = q.front();
            q.pop();
            if(node->data == a) start = node;
            if(node ->left){
                parent[node ->left] = node;
                q.push(node ->left);
            }
            if(node ->right){
                parent[node ->right] = node;
                q.push(node ->right);
            }
        }
        return start;
    }
    public:
    int findDist(Node* root, int a, int b) {
        unordered_map<Node *,Node *> parent;
        Node * start = fillParent(parent,root,a);
        queue<Node * >q;
        set<Node*>vis;
        vis.insert(start);
        q.push(start);
        int level = 0;
        while(!q.empty()){
            int sz = q.size();
            int flage = false; 
            for(int i=0; i<sz; i++){
                Node * node = q.front();
                q.pop();
                if(node->data == b){
                    flage = true;
                    break;
                }
                if(node->left && vis.find(node->left) == vis.end()){
                    q.push(node->left);
                    vis.insert(node->left);
                }
                if(node->right && vis.find(node->right) == vis.end()){
                    q.push(node->right);
                    vis.insert(node->right);
                }
                if(parent[node] && vis.find(parent[node]) == vis.end()){
                    q.push(parent[node]);
                    vis.insert(parent[node]);
                }
            }
            if(flage) break;
            level++;
        }
        return level;
    }
};

File: JAVA/Min distance between two given nodes of a Binary Tree.java
class GfG {
    Node cans(Node root,int a,int b)
    {
        if(root==null||root.data==a||root.data==b) return root;
        Node ln=cans(root.left,a,b);
         Node rn=cans(root.right,a,b);
         if(rn==null) return ln;
         if(ln==null) return rn;
         return root;
    }
    int dist(Node root,int a)
    {
        if(root==null) return -1;
        else if(root.data==a) return 0;
        int ld=dist(root.left,a);
        int rd=dist(root.right,a);
        if(ld>=0) return ld+1;
        return (rd>=0)?rd+1:-1;
        
    }
    int findDist(Node root, int a, int b)
    {
       Node nn=cans(root,a,b);
       int ld=dist(nn,a);
       int rd=dist(nn,b);
       return ld+rd;
    }
}
